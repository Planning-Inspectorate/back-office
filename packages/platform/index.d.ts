import './config/dotenv';
import './http/ttl';
import './util/CryptoUtils';
import './util/date';
import './util/filter';
import './util/url';
import './validators/date';
import './validators/postcode';

declare module '@pins/platform' {
	export * from './config/dotenv';
	export * from './http/ttl';
	export * from './util/CryptoUtils';
	export * from './util/date';
	export * from './util/filter';
	export * from './util/url';
	export * from './validators/date';
	export * from './validators/postcode';
}
