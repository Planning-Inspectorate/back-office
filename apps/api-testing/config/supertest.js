// @ts-nocheck
import supertest from 'supertest';
import config from './config.cjs';
import { Logger } from '../utils/logger.js';

const logger = new Logger();
logger.log('green', '************* Tests starting ****************');
logger.log('cyan', `Running tests against: ${config.baseUrl}`);
logger.log('green', '*********************************************');

const request = supertest(config.baseUrl);

export default request;
