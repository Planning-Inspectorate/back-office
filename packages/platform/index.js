export * from './http/hooks.js';
export * from './testing/assets/assets.js';
export * from './testing/fake.js';
export * from './testing/html-parser.js';
export * from './utils/date.js';
export * from './utils/dotenv.js';
export * from './utils/filter.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
