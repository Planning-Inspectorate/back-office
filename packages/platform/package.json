{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=16.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.0",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.1",
    "@types/lodash-es": "^4.17.6",
    "type-fest": "^2.12.2"
  }
}
