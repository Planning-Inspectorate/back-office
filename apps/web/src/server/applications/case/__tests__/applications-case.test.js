import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import { fixtureCases } from '../../../../../testing/applications/fixtures/cases.js';
import { createTestEnvironment } from '../../../../../testing/index.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);

const nocks = () => {
	nock('http://test/').get('/applications/case-team').reply(200, {});
	nock('http://test/').get('/applications/123').reply(200, fixtureCases[3]);
};

describe('Applications case pages', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	afterAll(() => {
		nock.cleanAll();
	});

	const baseUrl = '/applications-service/case/123';

	describe('Project overview page', () => {
		describe('GET /case/123', () => {
			beforeEach(async () => {
				nocks();
				await request.get('/applications-service/case-team');
			});

			it('should render the page', async () => {
				const response = await request.get(baseUrl);
				const element = parseHtml(response.text);

				expect(element.innerHTML).toMatchSnapshot();
				expect(element.innerHTML).toContain('Overview');
			});
		});
	});

	describe('Project information page', () => {
		describe('GET /case/123/project-information', () => {
			describe('When domainType is inspector', () => {
				beforeEach(async () => {
					nock('http://test/').get('/applications/inspector').reply(200, {});
					nock('http://test/').get('/applications/123').reply(200, fixtureCases[6]);

					await request.get('/applications-service/inspector');
				});

				it('should show no change links or publishing links', async () => {
					const response = await request.get(`${baseUrl}/project-information`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).not.toContain('Preview and publish project');
					expect(element.innerHTML).not.toContain('There are unpublished changes');

					// if user is inspector the link to examination timetable page should not appear
					expect(element.innerHTML).not.toContain('Examination timetable');
				});
			});

			describe('When domainType is not inspector', () => {
				beforeEach(async () => {
					await request.get('/applications-service/case-team');
				});

				describe('When the case is not published', () => {
					beforeEach(async () => {
						nocks();
					});

					it('should show no info about previous publishings', async () => {
						const response = await request.get(`${baseUrl}/project-information`);
						const element = parseHtml(response.text);

						expect(element.innerHTML).toMatchSnapshot();
						expect(element.innerHTML).toContain('Change');
						expect(element.innerHTML).not.toContain('There are unpublished changes');
						expect(element.innerHTML).toContain('Internal use only');
					});
				});

				describe('When the case is published', () => {
					beforeEach(async () => {
						nock('http://test/').get('/applications/case-team').reply(200, {});
					});

					it('with no pending changes, should show publishing info and unpublish link', async () => {
						nock('http://test/').get('/applications/123').reply(200, fixtureCases[5]);

						const response = await request.get(`${baseUrl}/project-information`);
						const element = parseHtml(response.text);

						expect(element.innerHTML).toMatchSnapshot();
						expect(element.innerHTML).not.toContain('There are unpublished changes');
					});

					it('with pending changes, should show publishing info, unpublish link and warning about unpublished changes', async () => {
						nock('http://test/').get('/applications/123').reply(200, fixtureCases[6]);

						const response = await request.get(`${baseUrl}/project-information`);
						const element = parseHtml(response.text);

						expect(element.innerHTML).toMatchSnapshot();
						expect(element.innerHTML).toContain('There are unpublished changes');
					});
				});
			});
		});
	});

	describe('Preview and publish page', () => {
		describe('GET /case/123/preview-and-publish', () => {
			describe('When domainType is inspector', () => {
				beforeEach(async () => {
					nock('http://test/').get('/applications/inspector').reply(200, {});
					await request.get('/applications-service/inspector');
				});

				it('should not render the page', async () => {
					const response = await request.get(`${baseUrl}/preview-and-publish`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).toContain('problem with your login');
				});
			});

			describe('When domainType is not inspector', () => {
				beforeEach(async () => {
					nock('http://test/').get('/applications/case-team').reply(200, {});
					await request.get('/applications-service/case-team');
				});

				it('if already published, should render the page with the previous publishing info', async () => {
					nock('http://test/').get('/applications/123').reply(200, fixtureCases[6]);

					const response = await request.get(`${baseUrl}/preview-and-publish`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).toContain('Accept and publish changes');
				});

				it('if not yet published, should render the page with no previous publishing info', async () => {
					nock('http://test/').get('/applications/123').reply(200, fixtureCases[3]);

					const response = await request.get(`${baseUrl}/preview-and-publish`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).toContain('Accept and publish project');
				});
			});
		});

		describe('POST /case/123/preview-and-publish', () => {
			describe('When domainType is inspector', () => {
				beforeEach(async () => {
					nock('http://test/').get('/applications/inspector').reply(200, {});
					await request.get('/applications-service/inspector');
				});

				it('should not render the page', async () => {
					const response = await request.post(`${baseUrl}/preview-and-publish`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).toContain('problem with your login');
				});
			});

			describe('When domainType is not inspector', () => {
				beforeEach(async () => {
					nock('http://test/').get('/applications/case-team').reply(200, {});
					nock('http://test/')
						.patch('/applications/123/publish')
						.reply(200, { publishedDate: 1_673_882_517 });
					await request.get('/applications-service/case-team');
				});

				it('if the case is not published yet, should go the first-time success banner page', async () => {
					nock('http://test/').get('/applications/123').reply(200, fixtureCases[3]);

					const response = await request.post(`${baseUrl}/preview-and-publish`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).toContain('Project page published ');
				});

				it('if the case is already published, should go the 2nd-time success banner page', async () => {
					nock('http://test/').get('/applications/123').reply(200, fixtureCases[6]);

					const response = await request.post(`${baseUrl}/preview-and-publish`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).toContain('Project page updates published');
				});
			});
		});
	});

	describe('Unpublish case page', () => {
		beforeEach(async () => {
			nock('http://test/').get('/applications/case-team').reply(200, {});
			nock('http://test/').get('/applications/123').reply(200, fixtureCases[6]);

			await request.get('/applications-service/case-team');
		});

		describe('GET /case/123/unpublish', () => {
			it('should render the page with the right info', async () => {
				const response = await request.get(`${baseUrl}/unpublish`);
				const element = parseHtml(response.text);

				expect(element.innerHTML).toMatchSnapshot();
				expect(element.innerHTML).toContain('Unpublish the project');
			});
		});

		describe('POST /case/123/unpublish', () => {
			it('should go to success page if unpublished worked', async () => {
				const response = await request.post(`${baseUrl}/unpublish`);

				expect(response?.headers?.location).toEqual('./successfully-unpublished');
			});
		});
	});
});
