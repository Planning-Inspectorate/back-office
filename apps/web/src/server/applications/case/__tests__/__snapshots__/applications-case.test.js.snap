// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applications view case summary GET /case should render the page 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <h1 class=\\"govuk-heading-xl\\">Page not found</h1>
            <p class=\\"govuk-body\\">If you typed the web address, check it is correct.</p>
            <p class=\\"govuk-body\\">If you pasted the web address, check you copied the entire address.</p>
            <p             class=\\"govuk-body\\"><a href=\\"https://has-appeal.herokuapp.com/help/contact\\" class=\\"govuk-link\\">Contact the Planning Inspectorate Customer Support</a> if
                the problem persists</p>
        </div>
    </div>
</main>"
`;
